### PR TITLE
chore: added useFocusTrap hook to focus inputs inside panel.

### PR DESCRIPTION
**Pull request by mfazil-eightfold**
## SUMMARY:
1) Added new hook - `useFocusTrap`
2) Enable `focusTrap` to trap focus inside the panel component

https://user-images.githubusercontent.com/109717425/193685278-4df035b5-34bd-4d78-a658-8e8885971460.mov

## JIRA TASK (Eightfold Employees Only):
https://eightfoldai.atlassian.net/browse/ENG-31266

## CHANGE TYPE:

-   [ ] Bugfix Pull Request
-   [x] Feature Pull Request

## TEST COVERAGE:

-   [ ] Tests for this change already exist
-   [ ] I have added unittests for this change

### CODE REVIEW

File: src/hooks/useFocusTrap.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Based on https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element
+import { useRef, useEffect } from 'react';
+
+const KEYCODE_TAB = 9;
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:4]: Should use https://github.com/EightfoldAI/octuple/blob/main/src/shared/utilities/eventKeys.ts

File: src/hooks/useFocusTrap.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export function useFocusTrap(visible = true) {
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:6]: const + type, e.g. export const useFocusTrap = (visible: boolean = true): **whatever return type elref is goes here** => {

File: src/hooks/useFocusTrap.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    const elRef = useRef(null);
+    const handleFocus = (e: KeyboardEvent) => {
+        var focusableEls = elRef.current.querySelectorAll(
+                'a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select'
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:10]: Need to consider any element with tab index of 0, input[type="range"], please look through Octuple and add any that might be missing.

File: src/hooks/useFocusTrap.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// Based on https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element
+import { useRef, useEffect } from 'react';
+
+const KEYCODE_TAB = 9;
+
+export function useFocusTrap(visible = true) {
+    const elRef = useRef(null);
+    const handleFocus = (e: KeyboardEvent) => {
+        var focusableEls = elRef.current.querySelectorAll(
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:9]: const and type

File: src/hooks/useFocusTrap.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                'a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select'
+            ),
+            firstFocusableEl = focusableEls[0],
+            lastFocusableEl = focusableEls[focusableEls.length - 1];
+
+        var isTabPressed = e.key === 'Tab' || e.keyCode === KEYCODE_TAB;
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:15]: const and type

File: src/hooks/useFocusTrap.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+const KEYCODE_TAB = 9;
+
+export function useFocusTrap(visible = true) {
+    const elRef = useRef(null);
+    const handleFocus = (e: KeyboardEvent) => {
+        var focusableEls = elRef.current.querySelectorAll(
+                'a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select'
+            ),
+            firstFocusableEl = focusableEls[0],
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:12]: type

File: src/hooks/useFocusTrap.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
+// Based on https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element
+import { useRef, useEffect } from 'react';
+
+const KEYCODE_TAB = 9;
+
+export function useFocusTrap(visible = true) {
+    const elRef = useRef(null);
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:7]: type

File: src/components/Panel/Panel.types.ts [hunk 8]
```diff
@@ -198,6 +198,11 @@ export interface PanelProps extends Omit<OcBaseProps<HTMLElement>, 'title'> {
      * @default true
      */
     scrollLock?: boolean;
+    /**
+     * Set this to trap focus inside panel
+     * @default false
+     */
+    focusTrap?: boolean;
```
Comment (dkilgore-eightfold) [inline src/components/Panel/Panel.types.ts:205]: alphabetize and add @default annotation.

File: src/components/Panel/Panel.tsx [hunk 9]
```diff
@@ -273,34 +275,38 @@ export const Panel = React.forwardRef<PanelRef, PanelProps>(
...
-                                        aria-hidden={!visible}
-                                    >
+                                    <div ref={focusRef}>
                                         <div
-                                            ref={panelRef}
-                                            className={panelClasses}
-                                            onClick={stopPropagation}
-                                            style={getPanelStyle()}
+                                            {...rest}
+                                            tabIndex={-1}
```
Comment (dkilgore-eightfold) [inline src/components/Panel/Panel.tsx:290]: Can this tabIndex of -1 be removed now that the focus trap exists??

File: src/components/Panel/Panel.tsx [hunk 10]
```diff
@@ -273,34 +275,38 @@ export const Panel = React.forwardRef<PanelRef, PanelProps>(
...
-                                        ref={containerRef}
-                                        className={panelBackdropClasses}
-                                        onClick={(
-                                            e: React.MouseEvent<HTMLDivElement>
-                                        ) => {
-                                            maskClosable && onClose(e);
-                                        }}
-                                        aria-hidden={!visible}
-                                    >
+                                    <div ref={focusRef}>
```
Comment (dkilgore-eightfold) [inline src/components/Panel/Panel.tsx:287]: This focusRef may need to move inside the container

File: src/hooks/useFocusTrap.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+            firstFocusableEl = focusableEls[0],
+            lastFocusableEl = focusableEls[focusableEls.length - 1];
+
+        var isTabPressed = e.key === 'Tab' || e.keyCode === KEYCODE_TAB;
+
+        if (!isTabPressed) {
+            return;
+        }
+
+        if (e.shiftKey) {
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:21]: please use https://github.com/EightfoldAI/octuple/blob/main/src/shared/utilities/eventKeys.ts

File: src/hooks/useFocusTrap.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
+// Based on https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element
```
Comment (dkilgore-eightfold) [inline src/hooks/useFocusTrap.ts:1]: Question: Doesn't floating-ui package have a focusTrap? I though it did, but cannot find it.

**Review by dkilgore-eightfold — CHANGES_REQUESTED** [review]

File: src/components/Panel/Panel.tsx [hunk 13]
```diff
@@ -273,34 +275,38 @@ export const Panel = React.forwardRef<PanelRef, PanelProps>(
 
         useImperativeHandle(ref, () => operations);
 
+        const focusRef = useFocusTrap(focusTrap && visible);
```
Comment (dkilgore-eightfold) [inline src/components/Panel/Panel.tsx:278]: type

File: src/hooks/useFocusTrap.ts (hunk 3 above)
Comment (dkilgore-eightfold) [reply src/hooks/useFocusTrap.ts:10]: perhaps something like this, but typed :) --> https://zellwk.com/blog/keyboard-focusable-elements/

**Review by dkilgore-eightfold — CHANGES_REQUESTED** [review]

File: src/hooks/useFocusTrap.ts (hunk 12 above)
Comment (mfazil-eightfold) [reply src/hooks/useFocusTrap.ts]: @ychhabra-eightfold can you help here?

**Review by dkilgore-eightfold — APPROVED** [review]

**Pull request merged**